src/Compilers/Core/Portable/Syntax/InternalSyntax/SyntaxListPool.cs

```diff
@@ -2,24 +2,43 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+// #define LOG
+
 using System;
+#if LOG
 using System.Collections.Generic;
 using System.Diagnostics;
-using Roslyn.Utilities;
+#endif
+using Microsoft.CodeAnalysis.PooledObjects;
 
 namespace Microsoft.CodeAnalysis.Syntax.InternalSyntax
 {
     internal class SyntaxListPool
     {
-        private ArrayElement<SyntaxListBuilder?>[] _freeList = new ArrayElement<SyntaxListBuilder?>[10];
+        private static readonly ObjectPool<SyntaxListPool> s_listPool = new ObjectPool<SyntaxListPool>(() => new SyntaxListPool());
+
+        private const int InitialFreeListSize = 16;
+        private const int InitialBuilderCapacity = 32;
```

Comment:
I played around with these numbers a bit and these seemed to work well for the Roslyn sln